### ISSUE TITLE
CS8602 false positive after TryGetValue

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#73274]
**Version Used**: VS 17.11 Preview 1 (34828.26.main)

**Steps to Reproduce**:

[Sharplab](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIpwCwG5lQDMmM6AwugN7Lq2bFTboCyAFAJRU108AiAlgGMALvwD2AOwCGAJwCeAHiyoANOgAy/AM7CFAIzFiANgD4T6YOgC86CQFMA7h3xIePA8fR3rFgHQAVeQBxO2EANSkjAFc7VgAiAAt+OLUxKOENbV0PUwB+dCMs9hc3On4AM3RWAEI7dm5S6ldSt0KdH3snYoa3AF8eniwATlY24V8yNIlhbubafqReoA===)

**Diagnostic Id**: CS8602

**Expected Behavior**:

No warning

**Actual Behavior**:

A warning that `list` is dereferenced when it may be null. When in fact it cannot be null because of the simple check and assignment I do.

The really interesting thing is this only shows up when targeting .NET 6 and 8 for me. .NET Standard 2.0 and 2.1 targets don't mis-fire here.
It looks like .NET 6 added `[MaybeNullWhen(false)]` to the `TryGetValue` output parameter. I bet that's what is triggering this, though it shouldn't, since I check the false condition and assign a non-null value.

**Comment by jjonescz** [comment]
```cs
Dictionary<string, List<bool>> d = new();
bool e = d.TryGetValue("hi", out List<bool>? list);
if (!e)
{
    list = new();
}
Console.WriteLine(list.Count);
```

Nullability info isn't propagated through boolean variables like that. See https://github.com/dotnet/roslyn/issues/27011.

If you "inline" the check, it works:

```cs
Dictionary<string, List<bool>> d = new();
if (!d.TryGetValue("hi", out List<bool>? list))
{
    list = new();
}
Console.WriteLine(list.Count);
```

**Comment by jjonescz** [comment]
I agree it would be useful to support this, but I think it needs a language proposal - https://github.com/dotnet/roslyn/issues/27011#issuecomment-481398819.

**Comment by jaredpar** [comment]
The current behavior is indeed "By Design". I agree it's unfortunate and it would be nice if this worked. Getting this to work though is not a trivial undertaking, it's a significant amount of design and implementation work. Rikki wrote up a [summary][summary] of the problems that this poses.

[summary]: https://github.com/dotnet/roslyn/issues/66693#issuecomment-1416871548
